### ISSUE TITLE
Shouldn't Request.Form be populated on JSON posts?

### ISSUE THREAD

**Issue by SaphuA** [aspnet/HttpAbstractions#491]
It is my understanding that, when doing a post to an action method, Request.Form is supposed to contain the submitted key/value pairs, and that this is supposed to support the JSON content type.

If that is not true then please ignore and close this.

I am testing this in a default asp.net mvc rc1 web application.

Doing an ajax post and then accessing Request.Form throws an exception with the message: *Incorrect Content-Type: application/json*

This is caused because the [FormFeature](https://github.com/aspnet/HttpAbstractions/blob/67aa2546a87f0617005e6f5571ed7fed8ab2f088/src/Microsoft.AspNet.Http/Features/FormFeature.cs#L31) is instantiated with an HttpRequest. So when calling [ReadForm](https://github.com/aspnet/HttpAbstractions/blob/67aa2546a87f0617005e6f5571ed7fed8ab2f088/src/Microsoft.AspNet.Http/Features/FormFeature.cs#L76) *Form* is *null* and *HasFormContentType* is *false*, because only the content types *"application/x-www-form-urlencoded"* and *"multipart/form-data"* are allowed, causing it to throw the exception.

Is my understanding of Request.Form wrong or is this by design?

**Comment by davidfowl** [comment]
It's by design. Form is for `application/x-www-form-urlencoded` not JSON posts. If you want to get the JSON body then get the request stream and deserialize it.

**Comment by bojingo** [comment]
There should be an easier way to "test" whether the Request.Form will have valid data or throw an InvalidOperationException.  Something like Request.HasFormData?  As a matter of fact, why throw an exception instead of just having an empty dictionary?  

I created "middleware" which was working great intercepting and processing form request arguments, until I added a request that processes Json and started getting InvalidOperationException that I have to now find a clean way to code around.
